libraries/chain/transaction_context.cpp

```diff
@@ -557,7 +557,9 @@ namespace eosio { namespace chain {
    {
       uint32_t new_action_ordinal = trace->action_traces.size() + 1;
 
-      trace->action_traces.reserve( new_action_ordinal );
+      if (trace->action_traces.capacity() < new_action_ordinal) {
+         trace->action_traces.reserve( std::max((size_t)new_action_ordinal, trace->action_traces.capacity() * 2 ));
+      }
```

Comment:
I think it would be be better to avoid reserve entirely.  You can copy provided_action and then move it into emplace_back.

Reply:
hi, I've added one line to reserve the size at the beginning of transaction as well. however, as inline_action or require_recipient can push back the action_trace list, we can't avoid reserve entirely.

Reply:
In develop a similar change for block we went with `deque` instead of `vector` to avoid the copy all together.

Reply:
i'm also thinking whether to use ```std::shared_ptr<action>``` instead of ```action``` in action_trace will help to improve performance further.